numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1931,6 +1923,419 @@ operand_different_than_broadcast: {
+        }
+        else {
+            op_is_buffered = 1;
+        }
+
+        npy_intp inner_stride;
+        npy_intp reduce_outer_stride;
+        if (op_is_buffered) {
+            npy_intp itemsize = NIT_DTYPES(iter)[iop]->elsize;
+            /*
+             * A buffered operand has a stride of itemsize unless we use
+             * reduce logic.  In that case, either the inner or outer stride
+             * is 0.
+             */
+            if (is_reduce_op) {
+                if (NAD_STRIDES(reduce_axisdata)[iop] == 0) {
+                    inner_stride = itemsize;
+                    reduce_outer_stride = 0;
+                }
+                else {
+                    inner_stride = 0;
+                    reduce_outer_stride = itemsize;
+                }
+            }
+            else {
+                if (NIT_OPITFLAGS(iter)[iop] & NPY_OP_ITFLAG_BUF_SINGLESTRIDE
+                        && NAD_STRIDES(axisdata)[iop] == 0
+                        && !avoid_zero_strides) {
+                    /* This op is always 0 strides, so even the buffer is that. */
+                    inner_stride = 0;
+                    reduce_outer_stride = 0;
+                }
+                else {
+                    /* normal buffered op */
+                    inner_stride = itemsize;
+                    reduce_outer_stride = itemsize * best_coresize;
+                }
+            }
+        }
+        else {
+            inner_stride = NAD_STRIDES(axisdata)[iop];
+            reduce_outer_stride = NAD_STRIDES(reduce_axisdata)[iop];
+        }
+
+        if (!using_reduce) {
+            /* invalidate for now, since we should not use it */
+            reduce_outer_stride = NPY_MIN_INTP;
+        }
+
+        NPY_IT_DBG_PRINT(
+            "Iterator: op=%d (buffered=%d, reduce=%d, single-stride=%d):\n"
+            "    inner stride: %zd\n"
+            "    reduce outer stride: %zd  (if iterator uses reduce)\n",
+            iop, op_is_buffered, is_reduce_op,
+            (NIT_OPITFLAGS(iter)[iop] & NPY_OP_ITFLAG_BUF_SINGLESTRIDE) != 0,
+            inner_stride, reduce_outer_stride);
+
+        NBF_STRIDES(bufferdata)[iop] = inner_stride;
+        NBF_REDUCE_OUTERSTRIDES(bufferdata)[iop] = reduce_outer_stride;
+
+        /* The actual reduce usage may have changed! */
+        if (is_reduce_op) {
+            NIT_OPITFLAGS(iter)[iop] |= NPY_OP_ITFLAG_REDUCE;
+        }
+        else {
+            NIT_OPITFLAGS(iter)[iop] &= ~NPY_OP_ITFLAG_REDUCE;
+        }
+
+        if (!op_is_buffered) {
+            NIT_OPITFLAGS(iter)[iop] |= NPY_OP_ITFLAG_BUFNEVER;
+        }
+        else {
+            iterator_must_buffer = 1;
+        }
+    }
+
+    /*
+     * If we buffer or do not have grow-inner, make sure that the size is
+     * below the maximum_size, but a multiple of the coresize.
+     */
+    if (iterator_must_buffer || !(itflags & NPY_ITFLAG_GROWINNER)) {
+        if (maximum_size < best_size) {
```

Comment:
Is it possible to hit this? I thought the above guaranteed that `best_size <= maximum_size`? (Note the equality: `maximum_size == best_size` should be fine, no?)

Reply:
The "best_size` is the `core_size * outer_dim_size` which can be larger than the above (we only break _after_ found).
If it is equal, the calculation is unnecessary guaranteed to return the same value.

(Maybe `best_size` could be slightly better named, hmmm.)

Reply:
Ah, yes, I got confused - the `size` constraint inside the loop is simply that one cannot usefully have a core size larger than the maximum.